src/control/probe.rs

```diff
@@ -0,0 +1,463 @@
+        let group_probe = unsafe { self.next().unwrap_unchecked() };
+
+        // SAFETY: The caller ensures the hash table contains this type of items.
+        let full = unsafe { group_probe.full::<T>() };
+        ProbeItems {
+            probe: self,
+            index: full.index,
+            iter: full.iter,
+            marker: PhantomData,
+        }
+    }
+
+    /// Finds a full slot.
+    ///
+    /// Although not a safety violation, this method *will* loop infinitely if
+    /// the table does not contain an empty slot. You have been warned.
+    ///
+    /// # Safety
+    ///
+    /// The hash table must contain the given type of items.
+    pub(crate) fn find_full<T>(mut self, mut f: impl FnMut(&T) -> bool) -> Option<usize> {
+        self.assert_at_least_one_empty();
+        loop {
+            // SAFETY: We always return an item from the iterator.
+            let group_probe = unsafe { self.next().unwrap_unchecked() };
+
+            // SAFETY: The caller ensures the table contains the given type of items.
+            match unsafe { group_probe.find_full(&mut f) } {
+                Some(index) => return Some(index),
+                None => {
+                    if group_probe.has_empty() {
+                        return None;
+                    }
+                }
+            }
+        }
+    }
+
+    /// Finds a full slot or an empty slot to insert something.
+    ///
+    /// Although not a safety violation, this method *will* loop infinitely if
+    /// the table does not contain an empty slot. You have been warned.
+    ///
+    /// # Safety
+    ///
+    /// The hash table must contain the given type of items.
+    pub(crate) fn find_full_or_empty<T>(
+        mut self,
+        mut f: impl FnMut(&T) -> bool,
+    ) -> Result<usize, usize> {
+        self.assert_at_least_one_empty();
+        loop {
+            // SAFETY: We always return an item from the iterator.
+            let group_probe = unsafe { self.next().unwrap_unchecked() };
+
+            // SAFETY: The caller ensures the table contains the given type of items.
+            match unsafe { group_probe.find_full_or_empty(&mut f) } {
+                Ok(index) => return Ok(index),
+                Err(Some(index)) => return Err(index),
+                Err(None) => (),
+            }
+        }
+    }
+}
+impl Iterator for Probe {
+    type Item = GroupProbe;
+    fn next(&mut self) -> Option<GroupProbe> {
+        if cfg!(debug_assertions) {
+            if self.iter > self.bucket_mask {
+                // We should have found an empty bucket by now and ended the probe.
+                panic!(
+                    "Went past end of probe sequence searching for hash {:x?} (tag {:x?}, index {:?}) in {:x?}",
+                    self.hash,
+                    self.hash_tag,
+                    (self.hash as usize & self.bucket_mask),
+                    self.control_slice()
+                );
+            }
+            self.iter += 1;
```

Comment:
This is the bug that made me think my code wasn't working at first: I was checking if `self.stride > self.bucket_mask`, when really, I should be checking if `self.stride > self.bucket_mask * Group::WIDTH`.

Except… that can overflow. So, since this is all for a debug assertion anyway, I decided to just add an extra field when debug assertions are enabled that tracks how many iterations have passed. That way, we can debug "infinite" loops, but we don't actually take the performance hit in release mode.